File: sCore/src/main/scala/com/github/zvreifnitz/score/concurrent/VisibilityBarrier.scala
package com.github.zvreifnitz.score.concurrent

import com.github.zvreifnitz.jcore.concurrent.{VisibilityBarrier => jVisibilityBarrier}

object VisibilityBarrier {
  def visible[T](instance: T): T = jVisibilityBarrier.makeVisible(instance)

  def visible[T](instance: => T): T = jVisibilityBarrier.makeVisible(instance)
}

File: jCore/src/main/java/com/github/zvreifnitz/jcore/concurrent/VisibilityBarrier.java
package com.github.zvreifnitz.jcore.concurrent;

public final class VisibilityBarrier<T> {
    private final T instance;

    public VisibilityBarrier(final T instance) {
        this.instance = instance;
    }

    public static <I> I makeVisible(final I instance) {
        final VisibilityBarrier<I> barrier = new VisibilityBarrier<>(instance);
        return barrier.getInstance();
    }

    public T getInstance() {
        return this.instance;
    }
}
